File: concepts.go
package main

import "fmt"

type Animal struct { // <1>
	age int // <3>
}

func NewAnimal(age int) Animal {
	a := Animal{}
	a.age = age
	return a
}

func (a Animal) SetAge(age int) { // <4>
	a.age = age
}

func (a Animal) GetAge() int {
	return a.age
}

type IMakeSound interface { // <5>
	makeSound()
}

type Dog struct {
	Animal // <2>
	colour string
}

func NewDog(age int, colour string) Dog {
	d := Dog{}
	d.age = age
	d.colour = colour
	return d
}

func (d Dog) SetColour(colour string) {
	d.colour = colour
}

func (d Dog) GetColour() string {
	return d.colour
}

func (d Dog) MakeSound() { // <6>
	fmt.Println("Wuff")
}

func main() { // <7>
	var laika = NewDog(13, "black")
	laika.MakeSound()
	fmt.Println(laika.GetAge())
	fmt.Println(laika.GetColour())
}

/*
output:
 Wuff
 13
 black
*/

File: concepts.ts
class Animal { // <1>
    private age: number; // <3>

    constructor(age: number) {
        this.age = age;
    }

    public setAge(age: number) { // <4>
        this.age = age;
    }

    public getAge(): number {
        return this.age;
    }
}

interface IMakeSound { // <5>
    makeSound(): void;
}

class Dog extends Animal implements IMakeSound{ // <2> <6>
    private colour: string;

    constructor(age: number, colour: string) {
        super(age);
        this.colour = colour;
    }

    public setColour(colour: string) {
        this.colour = colour;
    }

    public getColour(): string {
        return this.colour;
    }

    public makeSound(): void { // <7>
        console.log('Wuff');
    }
}

let laika = new Dog(13, "black"); // <8>
laika.makeSound();
console.log(laika.getAge());
console.log(laika.getColour());

/** 
 * output:
 *  Wuff
 *  13
 *  black
 */